src/provider/index.ts

```diff
@@ -1,132 +1,114 @@
+import 'reflect-metadata';
+
+import {RaiseIntentPayload, APITopic, OpenPayload, FindIntentPayload, FindIntentsByContextPayload, BroadcastPayload} from '../client/internal';
+import {AppIntent, IntentResolution, Application, Intent} from '../client/main';
+
+import {injectable, inject} from 'inversify';
+import {Inject} from './common/Injectables';
+import {AppDirectory} from './model/AppDirectory';
+import {Model, FindFilter} from './model/Model';
+import {ContextHandler} from './controller/ContextHandler';
+import {IntentHandler} from './controller/IntentHandler';
+import {APIHandler} from './APIHandler';
+import {Injector} from './common/Injector';
+
+interface API {
+    [APITopic.OPEN]: [OpenPayload, void];
+    [APITopic.FIND_INTENT]: [FindIntentPayload, AppIntent];
+    [APITopic.FIND_INTENTS_BY_CONTEXT]: [FindIntentsByContextPayload, AppIntent[]];
+    [APITopic.BROADCAST]: [BroadcastPayload, void];
+    [APITopic.RAISE_INTENT]: [RaiseIntentPayload, IntentResolution];
 }
 
-// Message definitions
-export interface OpenArgs {
-    name: AppName;
-    context?: Context;
-}
-export interface ResolveArgs {
-    intent: IntentType;
-    context?: Context;
-}
-export interface SelectorResultArgs {
-    handle: number;
-    success: boolean;
-
-    /**
-     * The application that was selected by the user.
-     *
-     * Only specified when success is true.
-     */
-    app?: Application;
-
-    /**
-     * The reason that an app wasn't selected.
-     *
-     * Only specified when success is false.
-     */
-    reason?: string;
-
-    /**
-     * Determines the future behaviour of this intent
-     */
-    defaultAction: DefaultAction;
+@injectable()
+export class Main {
+    private _config = null;
+
+    @inject(Inject.APP_DIRECTORY)   private _directory!: AppDirectory;
+    @inject(Inject.MODEL)           private _model!: Model;
+
+    @inject(Inject.CONTEXT_HANDLER) private _contexts!: ContextHandler;
+    @inject(Inject.INTENT_HANDLER)  private _intents!: IntentHandler;
+
+    private apiHandler: APIHandler<APITopic>;
+
+    constructor() {
+        this.apiHandler = new APIHandler();
+    }
+
+    public async register(): Promise<void> {
+        Object.assign(window, {
+            main: this,
+            config: this._config,
+            directory: this._directory,
+            model: this._model,
+            contexts: this._contexts,
+            intents: this._intents
+        });
+
+        // Wait for creation of any injected components that require async initialization
+        await Injector.initialized;
+
+        // Current API
+        this.apiHandler.registerListeners<API>({
+            [APITopic.OPEN]: this.onOpen.bind(this),
+            [APITopic.FIND_INTENT]: this.findIntent.bind(this),
+            [APITopic.FIND_INTENTS_BY_CONTEXT]: this.findIntentsByContext.bind(this),
+            [APITopic.BROADCAST]: this.broadcast.bind(this),
+            [APITopic.RAISE_INTENT]: this.raiseIntent.bind(this)
```

Comment:
FYI, have also just bumped the TypeScript version to latest. Since these `bind()` calls all returned `any`, it basically made all of the type-safety within `APIHandler` completely pointless. :-)

Reply:
Yay! Been wanting to do that for a while. Can you also do a quick check for any other places we're binding and make sure the types match up.